File: 0046-permutations/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.backtrack(nums, 0, [])
        return self.result

    def backtrack(self, nums, k, path):
        if k == len(nums):
            self.result.append(path[:])
            return
        for i in range(len(nums)):
            if nums[i] in path:
                continue
            path.append(nums[i])
            self.backtrack(nums, k+1,path)
            path.pop()

File: 0046-permutations/javascript/Solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let result = []
    let path = []
    let backtrack = k => {
        if (k == nums.length) {
            result.push([...path])
            return
        }
        for (let i = 0; i < nums.length; ++i) {
            if (path.includes(nums[i])) {
                continue
            }
            path.push(nums[i])
            backtrack(k+1)
            path.pop()
        }
    }
    backtrack(0)
    return result
};